### ISSUE TITLE
implement next plan

### ISSUE THREAD

**Issue by danregima** [ReZorg/nn.nn#7]
<div class="UserMessageLayout-module__messageContainer__IOjwG"><span class="UserProblemStatementMessage-module__messageMeta__y5RUo">You asked</span><div style="--max-height: 500px;"><div class="ExpandableMarkdown-module__overflowing__D2ZUL ExpandableMarkdown-module__expanded__H8bpC"><div class="markdown-body MarkdownRenderer-module__container__sJ8jf ExpandableMarkdown-module__content__Rn4oS UserProblemStatementMessage-module__messageText__Jng0H" data-copilot-markdown="true" style="--MarkdownRenderer_streaming-fade-duration: 750ms;"><p>Implement the plan:</p>
<h1>Implementation Plan: Four New PLingua Modules for <code node="[object Object]">lang/pli/</code></h1>
<h2>Context and conventions to follow</h2>
<p>The existing library establishes patterns every new module must follow:</p>
<ul>
<li node="[object Object]">One feature area per <code node="[object Object]">.pli</code> file, with <code node="[object Object]">@model&lt;membrane_computing&gt;</code> header, block-comment banner sections, parameterized <code node="[object Object]">@module name(args)</code> definitions, <code node="[object Object]">@initial</code> membrane setup, named <code node="[object Object]">@rules</code> groups, and a trailing <code node="[object Object]">@info</code> block listing name/version/depends/modules.</li>
<li node="[object Object]">Fixed-point integer arithmetic (values ×100), object tuples like <code node="[object Object]">input{i, v}</code>, <code node="[object Object]">weight{i, j, w}</code>, <code node="[object Object]">output{p, d, v}</code>; results emitted to parent via <code node="[object Object]">(obj, out)</code> communication rules.</li>
<li node="[object Object]">Randomness via <code node="[object Object]">random(lo, hi)</code> and probabilistic masking as in <code node="[object Object]">dropout.pli</code>; membrane division/dissolution macros already sketched in <code node="[object Object]">integration.pli</code> (<code node="[object Object]">plingua_compat</code>).</li>
<li node="[object Object]">Reuse existing building blocks: <code node="[object Object]">linear_layer</code>, <code node="[object Object]">softmax</code>/<code node="[object Object]">log_softmax</code>, <code node="[object Object]">layer_norm</code> (batch_normalization.pli), <code node="[object Object]">scaled_dot_product_attention</code> / <code node="[object Object]">multi_head_attention</code> / <code node="[object Object]">transformer_encoder_block</code> (attention.pli), <code node="[object Object]">lookup_table</code> (layers.pli), criteria (criteria.pli).</li>
<li node="[object Object]">Tests go in <code node="[object Object]">test_extensions.pli</code> style (<code node="[object Object]">@test</code> blocks with <code node="[object Object]">@setup</code>, <code node="[object Object]">@rules</code>, <code node="[object Object]">assert_equal</code>), imported files listed at top; structural validation via <code node="[object Object]">lang/pli/validate.sh</code>.</li>
</ul>
<h2>1. Graph neural network modules — new file <code node="[object Object]">gnn.pli</code></h2>
<ul>
<li node="[object Object]"><strong>Graph representation</strong>: model each graph node as a membrane <code node="[object Object]">[node{i} ...]</code> holding feature objects <code node="[object Object]">feat{d, v}</code>; encode edges as <code node="[object Object]">edge{i, j, w}</code> objects in the enclosing <code node="[object Object]">graph</code> membrane, so the membrane structure itself is the graph and message passing maps onto inter-membrane communication rules.</li>
<li node="[object Object]"><strong><code node="[object Object]">message_passing(num_nodes, feat_dim)</code> module</strong>: rules that, for each edge, send a copy of the source node's features to the target node (<code node="[object Object]">msg{j, d, v}</code> objects routed via edge objects), then aggregate incoming messages per node (sum/mean/max variants, mirroring the pairwise-reduction pattern used for <code node="[object Object]">score_term</code> in attention.pli).</li>
<li node="[object Object]"><strong><code node="[object Object]">gcn_layer(num_nodes, in_dim, out_dim)</code></strong>: graph convolution — aggregate neighbor features weighted by normalized edge weights (precompute degree objects <code node="[object Object]">deg{i}</code> in <code node="[object Object]">@initial</code>), then apply a shared <code node="[object Object]">linear_layer</code> projection and activation to the aggregated features. Include backward rules propagating gradients back along edges (reverse message passing) consistent with the simplified backprop style of <code node="[object Object]">nn.pli</code>.</li>
<li node="[object Object]"><strong><code node="[object Object]">graph_attention_layer(num_nodes, feat_dim)</code> (GAT)</strong>: compute per-edge attention coefficients from concatenated node features via a small learnable vector, normalize with the existing row-softmax pattern, and use them as message weights — directly reusing the softmax-with-row-max stabilization rules from attention.pli.</li>
<li node="[object Object]"><strong><code node="[object Object]">graph_readout(num_nodes, feat_dim)</code></strong>: global pooling (sum/mean/max) of node features into a graph-level embedding for graph classification, emitted <code node="[object Object]">out</code> to the parent.</li>
<li node="[object Object]">Document degree normalization and self-loops (add <code node="[object Object]">edge{i, i, 100}</code> in <code node="[object Object]">@initial</code>).</li>
</ul>
<h2>2. Neuroevolution via membrane division rules — new file <code node="[object Object]">neuroevolution.pli</code></h2>
<ul>
<li node="[object Object]"><strong>Core idea</strong>: use P-Systems' native membrane division (already wrapped as the <code node="[object Object]">divide</code> macro in <code node="[object Object]">integration.pli</code>) so that each candidate network is a membrane whose genome is its <code node="[object Object]">weight{...}</code> object multiset; division = reproduction, dissolution = selection death.</li>
<li node="[object Object]"><strong><code node="[object Object]">genome_membrane(genome_size)</code></strong>: a membrane holding <code node="[object Object]">gene{k, v}</code> objects plus a <code node="[object Object]">fitness{f}</code> object; helper rules to encode/decode a network's weights into genes.</li>
<li node="[object Object]"><strong><code node="[object Object]">mutate(rate, magnitude)</code></strong>: probabilistic evolution rules that, with probability <code node="[object Object]">rate</code> (using the <code node="[object Object]">random(0,100) &gt;= p</code> guard pattern from dropout.pli), perturb a gene by <code node="[object Object]">random(-magnitude, magnitude)</code>.</li>
<li node="[object Object]"><strong><code node="[object Object]">crossover</code></strong>: an antiport-style exchange rule between two sibling genome membranes swapping gene subsets keyed by index ranges.</li>
<li node="[object Object]"><strong><code node="[object Object]">divide_and_mutate</code></strong>: the signature rule — a membrane with <code node="[object Object]">fitness{f}</code> above a threshold divides into two daughter membranes (<code node="[object Object]">[genome{i}] -&gt; [genome{2i}] [genome{2i+1}]</code>), one copying genes verbatim (elitism) and one entering the mutation phase; below-threshold membranes fire a dissolution rule releasing a <code node="[object Object]">died{i}</code> marker.</li>
<li node="[object Object]"><strong><code node="[object Object]">tournament_selection(pop_size, k)</code></strong>: pairwise fitness comparison rules that keep the max (same reduction idiom as <code node="[object Object]">cmax_table</code>), tagging winners with <code node="[object Object]">survive{i}</code>.</li>
<li node="[object Object]"><strong><code node="[object Object]">neuroevolution_loop(pop_size, generations, mutation_rate)</code></strong>: orchestrator membrane cycling evaluate → select → divide/mutate phases with a <code node="[object Object]">generation{g}</code> counter and phase-token objects (the phase-token pattern used by <code node="[object Object]">simulation_harness</code> in integration.pli).</li>
<li node="[object Object]">Also add a <strong>NEAT-flavored structural mutation</strong>: rules that add a new neuron membrane inside a genome (membrane creation) or add/remove an <code node="[object Object]">edge</code>/<code node="[object Object]">weight</code> object, tying neuroevolution to the GNN edge representation for topology evolution.</li>
</ul>
<h2>3. Probabilistic P-Systems for Bayesian layers — new file <code node="[object Object]">bayesian.pli</code></h2>
<ul>
<li node="[object Object]"><strong>Model header</strong>: introduce this file as using probabilistic rule semantics (stochastic P-Systems); document that rule probabilities are expressed with the <code node="[object Object]">random(0,100)</code> guard convention where the simulator lacks native probabilistic rules.</li>
<li node="[object Object]"><strong><code node="[object Object]">gaussian_sampler(mean, std)</code></strong>: approximate normal sampling by summing several uniform <code node="[object Object]">random</code> draws (central-limit approximation, ×100 scaled); this is the primitive all other modules use.</li>
<li node="[object Object]"><strong><code node="[object Object]">bayes_linear_layer(input_size, output_size)</code></strong>: replace point weights with distribution parameters <code node="[object Object]">w_mu{i, j, m}</code> and <code node="[object Object]">w_rho{i, j, r}</code> (softplus of rho = std); a <code node="[object Object]">sample_weights</code> rule group draws <code node="[object Object]">weight{i, j, mu + std*eps}</code> per forward pass via the reparameterization trick, then reuses the standard <code node="[object Object]">linear_layer</code> forward rules. Store the drawn <code node="[object Object]">eps{i, j, e}</code> for the backward pass.</li>
<li node="[object Object]"><strong><code node="[object Object]">kl_divergence_regularizer(prior_std)</code></strong>: rules computing the closed-form KL between the weight posterior and a Gaussian prior, accumulating a <code node="[object Object]">kl_loss{v}</code> object added to the criterion loss (reuses the pairwise-sum reduction idiom; complements the existing <code node="[object Object]">kl_div_criterion</code>).</li>
<li node="[object Object]"><strong>Backward rules (Bayes-by-Backprop)</strong>: gradient rules updating both <code node="[object Object]">w_mu</code> (grad passes straight through) and <code node="[object Object]">w_rho</code> (grad scaled by stored <code node="[object Object]">eps</code> and softplus derivative), following the update pattern of <code node="[object Object]">momentum_adam_rmsprop.pli</code>.</li>
<li node="[object Object]"><strong><code node="[object Object]">mc_dropout_predict(num_samples)</code></strong>: Monte-Carlo prediction — run <code node="[object Object]">num_samples</code> stochastic forward passes tagged with sample index <code node="[object Object]">s</code> (maximally parallel, like <code node="[object Object]">batch_processor</code>), then reduce to predictive mean and variance objects <code node="[object Object]">pred_mean{i, v}</code> / <code node="[object Object]">pred_var{i, v}</code> for uncertainty estimation.</li>
<li node="[object Object]"><strong><code node="[object Object]">probabilistic_rule</code> compat macro</strong>: extend <code node="[object Object]">plingua_compat</code> in integration.pli (or define locally) with a <code node="[object Object]">@macro prob_rule(p, lhs, rhs, membrane)</code> wrapper so files can target simulators with native probabilistic P-System rule syntax.</li>
</ul>
<h2>4. Full transformer decoder with causal masking — new file <code node="[object Object]">transformer_decoder.pli</code></h2>
<ul>
<li node="[object Object]"><strong><code node="[object Object]">causal_mask(seq_len)</code></strong>: the key addition — a rule/guard that suppresses attention scores where key position <code node="[object Object]">j</code> &gt; query position <code node="[object Object]">i</code>. Implement inside masked attention as a <code node="[object Object]">where j &lt;= i</code> guard on the score-term rules, plus an explicit variant that sets <code node="[object Object]">score{i, j, NEG_INF}</code> (a large negative sentinel, e.g. −10000) for <code node="[object Object]">j &gt; i</code> so the existing row-softmax rules from attention.pli drive masked weights to ~0.</li>
<li node="[object Object]"><strong><code node="[object Object]">masked_self_attention(seq_len, d_k)</code></strong>: clone of <code node="[object Object]">scaled_dot_product_attention</code> with the causal guard applied to <code node="[object Object]">score_term</code> generation; document that under maximal parallelism only the lower-triangular score set is ever created.</li>
<li node="[object Object]"><strong><code node="[object Object]">cross_attention(tgt_len, src_len, d_k)</code></strong>: attention where Q comes from decoder positions and K/V from encoder outputs — parameterize the existing attention module over two sequence lengths and accept <code node="[object Object]">enc_out{p, d, v}</code> objects injected from the encoder membrane.</li>
<li node="[object Object]"><strong><code node="[object Object]">transformer_decoder_block(tgt_len, src_len, d_model, num_heads, d_ff)</code></strong>: three-sublayer membrane composition mirroring <code node="[object Object]">transformer_encoder_block</code>: masked multi-head self-attention → add&amp;norm → multi-head cross-attention over encoder memory → add&amp;norm → position-wise FFN → add&amp;norm, reusing <code node="[object Object]">layer_norm</code>, <code node="[object Object]">linear_layer</code>, <code node="[object Object]">relu_activation</code>, and the residual-save pattern (<code node="[object Object]">x_saved</code>/<code node="[object Object]">norm1_saved</code>) already used in attention.pli.</li>
<li node="[object Object]"><strong><code node="[object Object]">transformer_decoder(num_layers, ...)</code></strong>: stack of decoder-block membranes wired by output→input communication rules (same chaining style as <code node="[object Object]">seq2seq</code> in recurrent.pli).</li>
<li node="[object Object]"><strong><code node="[object Object]">full_transformer(...)</code></strong>: end-to-end module composing token <code node="[object Object]">lookup_table</code> embeddings + a new <code node="[object Object]">positional_encoding(seq_len, d_model)</code> module (sinusoidal values precomputed in <code node="[object Object]">@initial</code> as <code node="[object Object]">pos_enc{p, d, v}</code> objects added to embeddings), the existing <code node="[object Object]">transformer_encoder_block</code> stack, the new decoder stack, and a final linear + <code node="[object Object]">log_softmax</code> projection to vocabulary logits.</li>
<li node="[object Object]"><strong><code node="[object Object]">greedy_decode(max_len)</code></strong>: autoregressive generation loop — a step-token drives: run decoder on tokens so far, take argmax of last position's logits (pairwise-max reduction), append <code node="[object Object]">token{t+1, id}</code>, repeat until <code node="[object Object]">eos</code> object or <code node="[object Object]">max_len</code>; note that causal masking makes incremental decoding correct without re-masking.</li>
</ul>
<h2>5. Tests, validation, and documentation</h2>
<ul>
<li node="[object Object]"><strong><code node="[object Object]">test_extensions.pli</code></strong>: add <code node="[object Object]">@import</code> lines for the four new files and <code node="[object Object]">@test</code> blocks per module with hand-computed fixed-point expectations: GCN aggregation on a 3-node triangle graph; a mutation/division step producing two daughter genomes; Bayesian linear forward with fixed <code node="[object Object]">eps</code> and its KL value; causal mask verification (attn_weight{2,3} absent / ~0, row 1 attends only to itself) and one decoder-block residual check.</li>
<li node="[object Object]"><strong>Validation</strong>: run <code node="[object Object]">lang/pli/validate.sh</code> to confirm balanced comments/braces/brackets/parens in all new files (and with <code node="[object Object]">PLINGUA=</code> compiler if available).</li>
<li node="[object Object]"><strong>README.md updates</strong>: move the four items from "Potential further extensions" to "Completed extensions" with file names; add feature bullets, API-reference entries, and short usage examples for each new module; extend the Features list (GNN, neuroevolution, Bayesian/uncertainty, decoder/generation).</li>
<li node="[object Object]"><strong><code node="[object Object]">integration.pli</code></strong>: register the new modules in <code node="[object Object]">module_registry</code> and (if adopted) add the <code node="[object Object]">prob_rule</code> macro to <code node="[object Object]">plingua_compat</code>.</li>
<li node="[object Object]"><strong><code node="[object Object]">demo.pli</code> (optional)</strong>: add small demos — node classification on a toy graph, evolving XOR weights, uncertainty on a noisy regression point, and greedy decoding of a 4-token sequence.</li>
</ul>
<h2>Suggested implementation order</h2>
<ol>
<li node="[object Object]"><code node="[object Object]">transformer_decoder.pli</code> (mostly composition of existing attention machinery; causal mask is a small delta)</li>
<li node="[object Object]"><code node="[object Object]">gnn.pli</code> (new edge/message idioms but reuses reduction and linear patterns)</li>
<li node="[object Object]"><code node="[object Object]">bayesian.pli</code> (needs the Gaussian-sampling primitive and dual-parameter updates)</li>
<li node="[object Object]"><code node="[object Object]">neuroevolution.pli</code> (depends on division/dissolution semantics; can reuse GNN edges for topology mutation)</li>
<li node="[object Object]">Tests, validate.sh run, README/integration updates last.</li>
</ol></div></div><button data-component="Button" type="button" aria-expanded="true" class="prc-Button-ButtonBase-9n-Xk ExpandableMarkdown-module__toggle__NBxAX" data-loading="false" data-no-visuals="true" data-size="small" data-variant="invisible"><span data-component="buttonContent" data-align="center" class="prc-Button-ButtonContent-Iohp5"><span data-component="text" class="prc-Button-Label-FWkx3">Show less</span></span></button></div></div>
